File: lab3_PPOIS/OnlineStore/Date.cpp
#include "Date.h"

bool Date::IsLeapYear(int year) const {
	return (year % 4 == 0 && year % 100 != 0) || (year % 400 == 0);
}

int  Date::MaxDays(int year, int month) const {
	switch (month) {
	case 2:
		return IsLeapYear(year) ? 29 : 28;
	case 4: case 6: case 9: case 11:
		return 30;
	default:
		return 31;
	}
}

Date::Date(int day, int month, int year)
	:day(day), month(month), year(year) {
	allInDays = day + ((month - 1) * 30) + ((year - 1899) * 365);
}

void  Date::ShowDate() const {
	cout << day << "." << month << "." << year << endl;
}

bool  Date::IsValidDate() const {
	if (year < 1899 || year > 2024) return false;
	if (month < 1 || month > 12) return false;
	if (day < 1 || day > MaxDays(year, month)) return false;
	return true;
}

File: PPOIS/sem4/lab1/models/date.py
class Date:
    def __init__(self, day: int, month: int, year: int):
        if not self.__is_valid_date(day, month, year):
            raise ValueError(f"Invalid date: {day}.{month}.{year}")
        self.__day = day
        self.__month = month
        self.__year = year


    @staticmethod
    def __is_leap_year(year: int) -> bool:
        return year % 4 == 0 and year % 100 != 0 or year % 400 == 0


    def __max_days_in_month(self, month: int, year: int) -> int:
        if month == 2:
            return 29 if self.__is_leap_year(year) else 28
        elif month in (4, 6, 9, 11):
            return 30
        else:
            return 31


    def __str__(self):
        return f"{self.__day}.{self.__month}.{self.__year}"


    def __is_valid_date(self, day: int, month: int, year: int) -> bool:
        if not 1901 < year < 2025:
            return False
        if not 1 <= month <= 12:
            return False
        if not 1 <= day <= self.__max_days_in_month(month, year):
            return False
        return True